modepy/quadrature/finding.py

```diff
@@ -0,0 +1,422 @@
+of this software and associated documentation files (the "Software"), to deal
+in the Software without restriction, including without limitation the rights
+to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+copies of the Software, and to permit persons to whom the Software is
+furnished to do so, subject to the following conditions:
+
+The above copyright notice and this permission notice shall be included in
+all copies or substantial portions of the Software.
+
+THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+THE SOFTWARE.
+"""
+
+import operator
+from collections.abc import Callable, Sequence
+from dataclasses import dataclass
+from functools import reduce
+from typing import TypeAlias
+
+import numpy as np
+import numpy.linalg as la
+
+
+# FIXME: Better name?
+Integrand: TypeAlias = Callable[[np.ndarray], np.ndarray]
+
+
+@dataclass(frozen=True)
+class _ProductIntegrand:
+    functions: Sequence[Integrand]
+
+    def __call__(self, points: np.ndarray) -> np.ndarray:
+        return reduce(operator.mul, (f(points) for f in self.functions))
+
+
+@dataclass(frozen=True)
+class _ConjugateIntegrand:
+    function: Integrand
+
+    def __call__(self, points: np.ndarray) -> np.ndarray:
+        return self.function(points).conj()
+
+
+def _identity_integrand(points: np.ndarray) -> np.ndarray:
+    return points
+
+
+@dataclass(frozen=True)
+class LinearCombinationIntegrand:
+    """
+    .. note::
+
+        New linear combinations should be created by :func:`linearly_combine`,
+        which flattens nested combinations by taking advantage of associativity,
+        substantially decreasing evaluation cost.
+
+    .. autoattribute:: coefficients
+    .. autoattribute:: functions
+    .. automethod:: __call__
+    """
+    coefficients: np.ndarray
+    functions: Sequence[Integrand]
+
+    def __post_init__(self):
+        assert len(self.coefficients) == len(self.functions)
+
+    def __call__(self, points: np.ndarray) -> np.ndarray:
+        """Evaluate the linear combination of :attr:`functions` with
+        :attr:`coefficients` at *points*.
+        """
+        return sum(
+            (coeff * func(points)
+                for coeff, func in zip(self.coefficients, self.functions, strict=True)),
+            np.zeros(()))
+
+
+def linearly_combine(
+            coefficients: np.ndarray,
+            functions: Sequence[Integrand]
+        ) -> LinearCombinationIntegrand:
+    """
+    Takes advantage of associativity when linearly combining linear combinations.
```

Comment:
This description should be a bit more thorough? From what I can tell, flattening works only if each linear combination has a subset of the same functions and if they're all in the same order?

Reply:
Thanks! Looks good!